# Exercise 4.4 Data Consistency Checks_CLYC

## Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## Importing Data Sets

In [3]:
#Assign path
path = r'C:\Users\clany\OneDrive\Documents\CareerFoundry\Chap_4_Pythons Fundamentals for Data Analysts\Instacart Basket Analysis'

In [4]:
path

'C:\\Users\\clany\\OneDrive\\Documents\\CareerFoundry\\Chap_4_Pythons Fundamentals for Data Analysts\\Instacart Basket Analysis'

In [5]:
# Import ´products.csv´ data set from ´Original Data´ Folder as df_prods
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [8]:
# Import ´orders_wrangeld.csv´ Data set from ´Prepared Data´ folder as df_ords
df_ords = pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [9]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_last_order  
0                  8                    NaN  
1                  7                   15.0  
2                 12                   21.0  
3                  7                   29.0  
4                 15                   28.0

### Creating a small test dataframe

In [10]:
# Create a dataframe
df_test = pd.DataFrame()

In [11]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

## Mixed-Typed Data

In [13]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [14]:
df_test['mix'] = df_test['mix'].astype('str')

In [17]:
# Check for mixed types after changing to string
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

## Missing Values

In [19]:
# Finding missing values in df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [20]:
# View the above 16 values that are missing: create a new subset of df_prod called df_nan
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [21]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

### Creating a new dataframe without the missing rows

In [22]:
# First check the total number of rows
df_prods.shape

(49693, 5)

In [23]:
# Create a new dataframe df_prods_clean
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [24]:
# Check the new cleaned dataframe
df_prods_clean.shape

(49677, 5)

# Duplicates

In [25]:
# Locate duplicate rows in a new subset df_dup: only duplicate rows will be in it!
df_dup = df_prods_clean[df_prods_clean.duplicated()]

In [26]:
df_dup

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [27]:
# Check current number of rows
df_prods_clean.shape

(49677, 5)

In [28]:
# Delete the duplicate rows by creating a new datafram EXCLUDING the duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [29]:
df_prods_clean_no_dups.shape

(49672, 5)

## Export df_prods_clean_no_dups dataframe as “products_checked.csv” in the “Prepared Data” folder

In [30]:
# Export code to prepared_folder
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data','Prepared Data', 'products_checked.csv'))

# Task 4.5: Data Consistency Checks

#### Step 1: If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now

#### See above

#### Step 2: Run the df.describe() function on your df_ords dataframe. Share in a markdown cell whether anything about the data looks off or should be investigated further.

In [39]:
df_ords.describe()

order_id       user_id  number_of_orders  orders_day_of_week  \
count  3.421083e+06  3.421083e+06      3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05      1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04      1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00      1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04      5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05      1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05      2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05      1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

Would require further investigation: The max of number_of_orders shows a value of ´1´ however, the when checking the 25% which shows 5, the max value cannot be 1. 
Also, the min days_since_prior_order shows a value of 0 days which cannot be true.

#### Step 3: Check for mixed-type data in your df_ords dataframe

There were no mixed types found, since below function did not return ´mixed´

In [40]:
# Check for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### Step 4: If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

No mixed-type was found

In [42]:
df_ords.dtypes

order_id                    int64
user_id                     int64
eval_set                   object
number_of_orders            int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

#### Step 5: Run a check for missing values in your df_ords dataframe. Report your findings and propose an explanation for any missing values you find.

In [43]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
number_of_orders               0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [45]:
df_ords.head(20)

order_id  user_id eval_set  number_of_orders  orders_day_of_week  \
0    2539329        1    prior                 1                   2   
1    2398795        1    prior                 2                   3   
2     473747        1    prior                 3                   3   
3    2254736        1    prior                 4                   4   
4     431534        1    prior                 5                   4   
5    3367565        1    prior                 6                   2   
6     550135        1    prior                 7                   1   
7    3108588        1    prior                 8                   1   
8    2295261        1    prior                 9                   1   
9    2550362        1    prior                10                   4   
10   1187899        1    train                11                   4   
11   2168274        2    prior                 1                   2   
12   1501582        2    prior                 2                   5   
13   1901567        2    prior                 3                   1   
14    738281        2    prior                 4                   2   
15   1673511        2    prior                 5                   3   
16   1199898        2    prior                 6                   2   
17   3194192        2    prior                 7                   2   
18    788338        2    prior                 8                   1   
19   1718559        2    prior                 9                   2   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0  
11                 11                     NaN  
12                 10                    10.0  
13                 10                     3.0  
14                 10                     8.0  
15                 11                     8.0  
16                  9                    13.0  
17                 12                    14.0  
18                 15                    27.0  
19                  9                     8.0

In [47]:
# Creating a new  subset to see NaN values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [48]:
# View new subset with only NaN
df_ords_nan

order_id  user_id eval_set  number_of_orders  orders_day_of_week  \
0         2539329        1    prior                 1                   2   
11        2168274        2    prior                 1                   2   
26        1374495        3    prior                 1                   1   
39        3343014        4    prior                 1                   6   
45        2717275        5    prior                 1                   3   
...           ...      ...      ...               ...                 ...   
3420930    969311   206205    prior                 1                   4   
3420934   3189322   206206    prior                 1                   3   
3421002   2166133   206207    prior                 1                   6   
3421019   2227043   206208    prior                 1                   1   
3421069   3154581   206209    prior                 1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [51]:
# view last rows of df_ords to see the number of user_ids
df_ords.tail(10)

order_id  user_id eval_set  number_of_orders  orders_day_of_week  \
3421073   2307371   206209    prior                 5                   4   
3421074   3186442   206209    prior                 6                   0   
3421075    550836   206209    prior                 7                   2   
3421076   2129269   206209    prior                 8                   3   
3421077   2558525   206209    prior                 9                   4   
3421078   2266710   206209    prior                10                   5   
3421079   1854736   206209    prior                11                   4   
3421080    626363   206209    prior                12                   1   
3421081   2977660   206209    prior                13                   1   
3421082    272231   206209    train                14                   6   

         order_hour_of_day  days_since_prior_order  
3421073                 15                     3.0  
3421074                 16                     3.0  
3421075                 13                     9.0  
3421076                 17                    22.0  
3421077                 15                    22.0  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0

##### Result of the investigation of the reported NaN values: there are 206209 invalid values, the amount of user_ids ie customers is as well 206209. 
##### Each new user_id cannot have a prior order before the 1st order. Hence this provides a logical explaination of the absence of values at first order of a new customer. 

#### Step 6: Address the missing values using an appropriate method. Explain why you used your method of choice

##### Given the fact that the absence of the values align with the data regarding initial orders, offering insights into the customers
##### and their purchasing behavior, this column holds significance with the NaN included. I believe that there is no need to  take any further
##### steps to eliminate or amend the NaN data.

#### Step 7: Run a check for duplicate values in your df_ords data. Report your findings and propose an explanation.

In [52]:
# Check duplicate within a new subset df_ords_dups
df_ords_dups = df_ords[df_ords.duplicated()]

In [54]:
# View the new subset for duplicates
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, number_of_orders, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [55]:
# Check the dimension of new subset if any
df_ords_dups.shape

(0, 7)

#### Step 8: Address the duplicates using an appropriate method.
##### No duplicates found

#### Step 9: Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names

In [56]:
# Export df_ords as orders_checked
df_ords.to_csv(os.path.join(path, 'Data','Prepared Data', 'orders_checked.csv'))

In [57]:
# Export df_prods_clean_no_dups as products_checked (done already during above exercise)
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data','Prepared Data', 'products_checked.csv'))